zerver/webhooks/patreon/view.py

```diff
@@ -0,0 +1,123 @@
+from typing import Callable, Dict
+
+from django.http import HttpRequest, HttpResponse
+
+from zerver.decorator import webhook_view
+from zerver.lib.exceptions import UnsupportedWebhookEventType
+from zerver.lib.request import REQ, has_request_variables
+from zerver.lib.response import json_success
+from zerver.lib.validator import WildValue, check_int, check_string, to_wild_value
+from zerver.lib.webhooks.common import (
+    check_send_webhook_message,
+    validate_extract_webhook_http_header,
+)
+from zerver.models import UserProfile
+
+
+# The events for this integration contain the ":" character, which is not appropriate in a
+# filename and requires us to deviate from the common `get_http_headers_from_filename` method
+# from zerver.lib.webhooks.common.
+def get_custom_http_headers_from_filename(http_header_key: str) -> Callable[[str], Dict[str, str]]:
+    def fixture_to_headers(filename: str) -> Dict[str, str]:
+        event_type = ":".join(filename.split("_"))
+        return {http_header_key: event_type}
+
+    return fixture_to_headers
+
+
+fixture_to_headers = get_custom_http_headers_from_filename("HTTP_X_PATREON_EVENT")
+
+
+def handle_members_create_event(payload: WildValue) -> str:
+    included = payload["included"][0]
+    patron_count = included["attributes"]["patron_count"].tame(check_int)
+    user_data = payload["data"]["attributes"]
+    user_name = user_data["full_name"].tame(check_string)
+    body = f"{user_name} has joined as a member! :tada:\nYou now have {patron_count} patron(s)."
+
+    return body
+
+
+def handle_members_update_event(payload: WildValue) -> str:
+    included = payload["included"][0]
+    patron_count = included["attributes"]["patron_count"].tame(check_int)
+    user_data = payload["data"]["attributes"]
+    user_name = user_data["full_name"].tame(check_string)
+    body = (
+        f"{user_name} just updated their membership. :gear:\nYou now have {patron_count} patron(s)."
+    )
+
+    return body
+
+
+def handle_members_delete_event(payload: WildValue) -> str:
+    included = payload["included"][0]
+    patron_count = included["attributes"]["patron_count"].tame(check_int)
+    user_data = payload["data"]["attributes"]
+    user_name = user_data["full_name"].tame(check_string)
+    body = f"{user_name} just deleted their membership. :cross_mark:\nYou now have {patron_count} patron(s)."
+
+    return body
+
+
+def handle_members_pledge_create_event(payload: WildValue) -> str:
+    included = payload["included"][0]
+    patron_count = included["attributes"]["patron_count"].tame(check_int)
+    user_data = payload["data"]["attributes"]
+    user_name = user_data["full_name"].tame(check_string)
+    pledge_amount = user_data["pledge_amount_cents"].tame(check_int) / 100
+    body = f"{user_name} has created a new member pledge of ${pledge_amount}. :tada:\nYou now have {patron_count} patron(s)."
```

Comment:
I think it'd be nice if the ending sentence was always something like "You now have {patron_count} patrons donating a total of {dollars} per month." -- if we have the data for that.

(Would it make sense to just make this second sentence generated by a function, that can just be called from each of these, rather than duplicating it for each event type?)

Is `pledge_sum` the right value?

Reply:
> I think it'd be nice if the ending sentence was always something like "You now have {patron_count} patrons donating a total of {dollars} per month." -- if we have the data for that.
Is `pledge_sum` the right value?

I think that data exists within the webhook, but I'm not sure how to verify; their documentation lacks details on what information is being provided.

> (Would it make sense to just make this second sentence generated by a function, that can just be called from each of these, rather than duplicating it for each event type?)

Yes, sounds apt.